### PR TITLE
[1/N]Separate runtime env agent from dashboard agent

### PR DESCRIPTION
**Pull request by MisterLin1995**
## Why are these changes needed?

These changes are part of the work to achieve the goal of separating runtime env agent from dashboard agent described in #30528. This PR does the following the things:
1. Move runtime env agent code files to ray/python/_private/runtime_env/runtime_env_agent
2. Add a runtime_env_agent_manager in raylet's node manager to start and monitor runtime env agent process.
3. Make raylet restart dashboard agent process when it dies
4. Keep raylet fate-share with runtime env agent process
5. UT changes to match the new behavior
6. Other works to start / stop runtime env agent processes when using ray cluster
7. Other works to pass runtime env agent grpc port to different part of ray (dashboard / state api / proxy server)

## Related issue number

 #30528

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by iycheng** [comment]
Thanks @MisterLin1995 this PR is giant. I'll find some time next week reviewing this and might need to schedule a sync meeting for how to get it into the code base. Maybe we'll need to split it in someway.

**Review by fishbone — CHANGES_REQUESTED** [review]
Thanks for the work. I reviewed in high level and feel the goal of the work is clean, but just too big and we should think about how to deliver it incrementally.

Besides that, I do have some architecture questions.

- We should keep runtime env agent only visible to the raylet. Other module shouldn't see that. it might see other modules, though.
- Could the registration be more simple? If it's a module in raylet, maybe we should only pass the path of the main to raylet and raylet should be able to construct the cmd by itself to reduce redundant informations.

As for split, we probably can
- Refactor the agent manager and make it manage the dashboard agent.
- Move runtime env agent to the new place and make the old dashboard agent use that. In the mean time, make the runtime env agent able to start.
- Write runtime env agent manager in cpp and make it able to start the runtime env agent.
- Integrate it with raylet to replace the old ones and don't start runtime env agent in the old one by default.

File: dashboard/datacenter.py [hunk 1]
```diff
@@ -208,11 +208,16 @@ async def get_all_node_details(cls):
...
+        for node_id, (
+            http_port,
+            grpc_port,
+            runtime_env_grpc_port,
+        ) in DataSource.agents.items():
             agent_infos[node_id] = dict(
                 ipAddress=DataSource.node_id_to_ip[node_id],
                 httpPort=int(http_port or -1),
                 grpcPort=int(grpc_port or -1),
+                runtimeEnvGrpcPort=int(runtime_env_grpc_port),
```
[the remark this answers is not in the document]
Comment (MisterLin1995) [reply dashboard/datacenter.py:220]: Well, it's just the runtime env agent is directly called by other components other than Raylet, such as the client-server and state manager. Maybe we also need to provide RPC APIs in Raylet for these components to hide runtime env agent from them.

**Comment by MisterLin1995** [comment]
> * Refactor the agent manager and make it manage the dashboard agent.
I'm not quite sure about what's involved in this part, since it's already managing the dashboard agent.

> In the mean time, make the runtime env agent able to start.
Does this mean making the moved runtime env can be start both as module and as a separate process?

File: dashboard/datacenter.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (iycheng) [reply dashboard/datacenter.py:220]: Hmmm, I see. This is pretty interesting. Do you know where is the entry point for that?

**Comment by iycheng** [comment]
> > * Refactor the agent manager and make it manage the dashboard agent.
> 
> I'm not quite sure about what's involved in this part, since it's already managing the dashboard agent.
> 

I saw you update `class AgentManager : public rpc::AgentManagerServiceHandler {`. I think maybe we just have a PR make this manager more general? It's supposed to manage runtime env agent in the future I think?


> > In the mean time, make the runtime env agent able to start.
> 
> Does this mean making the moved runtime env can be start both as module and as a separate process?

I'm thinking that, the implementation is rpc unrelated, the dashboard agent's code is to expose the implementation and the new runtime env agent is also doing the same thing. This I mean the later one.

File: src/ray/raylet/agent_manager.cc [hunk 2]
```diff
@@ -131,7 +103,10 @@ void AgentManager::StartAgent() {
     RAY_LOG(INFO) << "Agent process with id " << agent_id << " exited, exit code "
                   << exit_code << ". ip " << reported_agent_ip_address_ << ". id "
                   << reported_agent_id_;
-
+    if (restart_when_agent_die_) {
+      delay_executor_([this]() { StartAgent(); }, /*ms*/ 10);
```
Comment (iycheng) [inline src/ray/raylet/agent_manager.cc:107]: Please log here too.

File: src/ray/protobuf/runtime_env_agent_manager.proto [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+enum RuntimeEnvAgentRpcStatus {
+  // OK.
+  RUNTIME_ENV_AGENT_RPC_STATUS_OK = 0;
+  // Failed.
+  RUNTIME_ENV_AGENT_RPC_STATUS_FAILED = 1;
+}
+
+message RegisterRuntimeEnvAgentRequest {
+  int32 agent_id = 1;
+  int32 agent_port = 2;
```
Comment (iycheng) [inline src/ray/protobuf/runtime_env_agent_manager.proto:29]: Do we use 0 for the port?

File: src/ray/protobuf/runtime_env_agent_manager.proto [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  string agent_ip_address = 3;
+}
+
+message RegisterRuntimeEnvAgentReply {
+  RuntimeEnvAgentRpcStatus status = 1;
+}
+
+service RuntimeEnvAgentManagerService {
```
Comment (iycheng) [inline src/ray/protobuf/runtime_env_agent_manager.proto:37]: is it possible for the runtime env agent just use the same protobuf as the dashboard agent

**Pull request closed without merging**